pandas/tests/io/test_excel.py

```diff
@@ -245,6 +245,23 @@ def test_index_col_empty(self, ext):
                                               names=["A", "B", "C"]))
         tm.assert_frame_equal(result, expected)
 
+    @pytest.mark.parametrize("index_col", [None, 2])
+    def test_index_col_with_unnamed(self, ext, index_col):
+        # see gh-18792
+        result = self.get_exceldf("test1", ext, "Sheet4",
+                                  index_col=index_col)
+        expected = DataFrame([["i1", "a", "x"], ["i2", "b", "y"]],
+                             columns=["Unnamed: 0", "col1", "col2"])
+
+        if index_col:
+            expected = expected.set_index(expected.columns[index_col])
+
+        # Python 2.x somehow interprets column
```

Comment:
are they somehow unicode / mixed strings? something is wrong here

Reply:
They are all unicode strings.  Python 3.x correctly sees them as string.

Reply:
and can you show the actual values and the type of each value.

Reply:
~~~python
from pandas import read_excel

filename = "pandas/tests/io/data/test1.xlsx"
df = read_excel(filename, "Sheet4")

print(df)
print(df.columns)
print(df.columns.inferred_type)

for col in df.columns:
   print(col, type(col))
~~~

Python 2:

~~~
  Unnamed: 0 col1 col2
0         i1    a    x
1         i2    b    y

Index([u'Unnamed: 0', u'col1', u'col2'], dtype='object')

'mixed'

('Unnamed: 0', <type 'str'>)
(u'col1', <type 'unicode'>)
(u'col2', <type 'unicode'>)
~~~

Python 3:

~~~
  Unnamed: 0 col1 col2
0         i1    a    x
1         i2    b    y

Index([u'Unnamed: 0', u'col1', u'col2'], dtype='object')

'string'

('Unnamed: 0', <type 'str'>)
(u'col1', <type 'str'>)
(u'col2', <type 'str'>)
~~~

Ah, so I was mistaken.  The columns are all string-types, but somehow Python 2.x gets a little confused over the dtype of `unicode` and `str` being a mixed-dtype.

Reply:
right so i think the index conversion is wrong somehow

the unnamed need to be unicode i think

Reply:
I looked at this again, and it actually looks like something went wrong with how I saved the file to Excel.  Let me try saving the file again.

Reply:
@jreback : Interesting, there are several tests in `test_excel` that actually have expected this behavior from `PY2` for quite some time now...

It seems that the columns in the Excel file are read in as `unicode` but are not cast to `string`.  All placeholder columns are of course `string`, leading to the `dtype` mismatch.

Reply:
I'm wondering if we really want to patch this discrepancy for Python 2.x, given that it for intensive purposes has not been extremely consequential, has existed in the codebase **DELIBERATELY** for ~3 years (based on the `git blame`), and support dropping soon.

Thoughts?

Reply:
no this needs to be fixed for 0.24.0 I think. the issue is that the index conversions are not creating unicode, rather they are creating strings

Reply:
> no this needs to be fixed for 0.24.0 I think. the issue is that the index conversions are not creating unicode, rather they are creating strings

No, that's not the problem.  If you read the data via CSV, the columns are all strings.

Reply:
it seems that when you add a Unnamed column, the 'Unnamed' should be wrapped in ``u()`` ?

Reply:
Adding "Unnamed" is part of parsing for all IO files.  If we unicoded column names, that would break CSV, which expects strings.  That's why I would prefer to go the other way and turn to string since that would be least impactful.

Keep in mind that this distinction will go away once we are Python 3.x-only.

Reply:
ok, why / where do you think things are being coerced to unicode? (and only for excel)?

its true that this is a PY2 only issue, but PY2 will still be around for a while, might as well get this right.

Reply:
> its true that this is a PY2 only issue, but PY2 will still be around for a while, might as well get this right.

Fair enough.

> why / where do you think things are being coerced to unicode? (and only for excel)?

Reply:
> its true that this is a PY2 only issue, but PY2 will still be around for a while, might as well get this right.

Fair enough.

> why / where do you think things are being coerced to unicode? (and only for excel)?

The reason AFAICT is that the Excel reader is giving us `unicode` in the first place.  This is something that we don't have control over, so we would have to convert to string ourselves before passing this along to our parsing logic.

If we want to patch this by converting everything to string, then I know where we can (relatively easily) do this.